### PR TITLE
Fkyd/xdma snax integration

### PR DESCRIPTION
**Pull request by IveanEx**
This PR enables the availability of the **_xdma_** on snax cluster, which will be the next generation distributed DMA focusing on intra-cluster data movement, inter-cluster data movement, and data movement between any cluster and global memory. 

The function of it includes: 
1. The data fetch unit that tries to fetch data from TCDM in 8B granularity
2. The datapath that, if one of the channels to TCDM is blocked by contention, other channels can still fetch the data It maximizes the bus utilization. 
3. The address generation unit that supports n-dimensional affine space fetching
4. The xdma datapath extension interface allows users to plug in any element-wise processing unit, enabling computaion-on-copy capability. 
5. Designed in Chisel, with configurable data extension, bit width, and many more. 

At present, only the intra-cluster capacity is enabled. Three extensions are designed as the reference of xdma datapath extension, including: 

1. Memset: Setting each byte in the given memory region to arbitrary value very fast, in one xdma request. 
2. Transposer: Transpose the 8x8 INT8 matrix on copy. 
3. MaxPool SIMD

The runtime is also provided to control xdma by standard snax CSR interface, with C headers automatically generated with the generation of hardware.

### CODE REVIEW

File: hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl [hunk 1]
```diff
@@ -392,6 +393,39 @@ for i in range(len(cfg['cores'])):
...
+          'snax_acc_name': xdma_instance_name,
+          'snax_tcdm_ports': snax_tcdm_ports,
+          'snax_tcdm_offset_start': tcdm_offset_start,
+          'snax_tcdm_offset_stop': tcdm_offset_stop
+        }
+    tcdm_offset_start += snax_tcdm_ports
+    prefix_snax_count += 1
+    total_snax_narrow_ports += snax_narrow_tcdm_ports
+    total_snax_wide_ports += snax_wide_tcdm_ports
+
```
Comment (rgantonio) [inline hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl:428]: I find this a bit odd, so when there is an accelerator, we won't use the `snax_xdma` ? I thought we use accelerators AND the xDMA too right? 

Was the intent of the SNAX xDMA to be a stand-alone accelerator? Because that's what the logic in this part says. The first condition `if ('snax_acc_cfg' in cfg['cores'][i]):` holds true then the SNAX xDMA won't be generated.

File: hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl [hunk 2]
```diff
@@ -826,6 +861,58 @@ total_snax_tcdm_ports = total_snax_narrow_ports + total_snax_wide_ports
 
       % endfor
     % endif
+  
+  % elif snax_core_acc[idx_key]['snax_xdma_flag']:
+    
+
+    % for jdx, jdx_key in enumerate(snax_core_acc[idx_key]['snax_acc_dict']):
+
+  // Accelerators controlled with custom instruction format ports
```
Comment (rgantonio) [inline hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl:870]: xDMA is not custom instruction controlled right?

File: hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl [hunk 3]
```diff
@@ -826,6 +861,58 @@ total_snax_tcdm_ports = total_snax_narrow_ports + total_snax_wide_ports
...
+    .csr_rsp_ready_i ( snax_csr_rsp_ready[${idx}] ),
+    //-----------------------------
+    // Hardware barrier is not supported by xdma at the moment
+    //-----------------------------
+    //-----------------------------
+    // TCDM ports
+    //-----------------------------
+    .tcdm_req_o  ( snax_tcdm_req[${snax_core_acc[idx_key]['snax_acc_dict'][jdx_key]['snax_tcdm_offset_stop']}:${snax_core_acc[idx_key]['snax_acc_dict'][jdx_key]['snax_tcdm_offset_start']}] ),
+    .tcdm_rsp_i  ( snax_tcdm_rsp[${snax_core_acc[idx_key]['snax_acc_dict'][jdx_key]['snax_tcdm_offset_stop']}:${snax_core_acc[idx_key]['snax_acc_dict'][jdx_key]['snax_tcdm_offset_start']}] )
+  );
```
Comment (rgantonio) [inline hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl:905]: Why does it have to be separated from the original accelerator generation? The ports still look the same!

If the point is to re-use the name you can always set the `snax_acc_name` to the xDMA name right?

File: hw/templates/snax_xdma_wrapper.sv.tpl
Comment (rgantonio) [inline hw/templates/snax_xdma_wrapper.sv.tpl:1]: I think this wrapper needs to be inside of SNAX cluster rather than outside like that of the accelerators 😄 

It makes much more sense that it is inside than outside because it is part of the cluster itself.

File: target/snitch_cluster/Makefile [hunk 4]
```diff
@@ -190,6 +190,27 @@ $(eval $(call generate_snax_gen,snax_streamer_gemmX))
...
+ifeq (${CFG_OVERRIDE}, cfg/snax-streamer-gemmX-xdma.hjson)
+
+	CLUSTER_NAME = snax_streamer_gemmX_xdma_cluster
+
+$(eval $(call generate_snax_gen,snax_streamer_gemmX))
+
+	SNAX_GEMMX_ROOT ?= $(ROOT)/hw/chisel_acc
+    include $(SNAX_GEMMX_ROOT)/Makefile
+
+# This is a temporary solution. Since the xdma is now generated within wrapperergen, we need to tell Makefile about this by adding these two files to the SNAX_GEN list
```
Comment (rgantonio) [inline target/snitch_cluster/Makefile:202]: Why is it temporary? It seems correct though. But also needs to be inside indeed! We can make it mandatory on the top!

File: target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    uint8_t *tcdm_16 = tcdm_baseaddress + 0x4000 * sizeof(uint8_t);
+    uint8_t *tcdm_32 = tcdm_baseaddress + 0x8000 * sizeof(uint8_t);
+    uint8_t *tcdm_48 = tcdm_baseaddress + 0xc000 * sizeof(uint8_t);
+    uint8_t *tcdm_64 = tcdm_baseaddress + 0x10000 * sizeof(uint8_t);
+    uint8_t *tcdm_80 = tcdm_baseaddress + 0x14000 * sizeof(uint8_t);
+    uint8_t *tcdm_96 = tcdm_baseaddress + 0x18000 * sizeof(uint8_t);
+    uint8_t *tcdm_112 = tcdm_baseaddress + 0x1c000 * sizeof(uint8_t);
+
+    // Using xdma core only
+    if (snrt_cluster_core_idx() == snrt_cluster_compute_core_num() - 1) {
```
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c:27]: Any particular reason why it has to be this long condition when you could just specify the actual core number where the xDMA is positioned? Or are you foreseeing that the xDMA will be placed in the 2nd to the last core always? However, that's not true right? Since we want the xDMA to be part of the accelerator cores too!

File: target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            printf("Error in disabling xdma extension 1\n");
+        } else {
+            printf("The xdma extension 1 is disabled\n");
+        }
+
+        if (xdma_disable_dst_ext(2) != 0) {
+            printf("Error in disabling xdma extension 2\n");
+        } else {
+            printf("The xdma extension 2 is disabled\n");
+        }
```
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c:52]: Nice sanity checkers! However, take note that sometimes this printing is not necessary as part of CI tests. I guess it's fine for now since we're not measuring performance here but rather functionality.

File: target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    uint8_t *tcdm_64 = tcdm_baseaddress + 0x10000 * sizeof(uint8_t);
+    uint8_t *tcdm_80 = tcdm_baseaddress + 0x14000 * sizeof(uint8_t);
+    uint8_t *tcdm_96 = tcdm_baseaddress + 0x18000 * sizeof(uint8_t);
+    uint8_t *tcdm_112 = tcdm_baseaddress + 0x1c000 * sizeof(uint8_t);
+
+    // Using xdma core only
+    if (snrt_cluster_core_idx() == snrt_cluster_compute_core_num() - 1) {
+        // The xdma core is the last compute core in the cluster
+        if (xdma_memcpy_1d(tcdm_0, tcdm_0, 0x4000 * sizeof(uint8_t)) != 0) {
+            printf("Error in xdma agu configuration\n");
```
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c:30]: Also when an error occurs, don't you think we do the put the return here immediately too?

File: target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+            "is %d\n",
+            task_id);
+        xdma_wait(task_id);
+
+        printf("The xdma is finished\n");
+        // Check the data
+        for (int i = 0; i < 0x4000; i++) {
+            if (tcdm_0[i] != 0xFF) {
+                printf("Error in memset\n");
+                err = 1;
```
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c:66]: I would increment the error value: `err++;` for every incorrect answer so we have information feedback on how much error is occurring. I mean when this loop breaks it will still do the `printf("The memset of 0KB - 16KB is correct\n");` right? Then you can also return without proceeding to the next to terminate simulation quick in case something is wrong.

File: target/snitch_cluster/sw/snax/xdma/include/snax-xdma-csr-addr.h [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+#define XDMA_DST_BOUND_PTR XDMA_DST_ADDR_PTR_MSB + 1
+#define XDMA_DST_STRIDE_PTR XDMA_DST_BOUND_PTR + XDMA_DST_DIM
+#define XDMA_DST_EXT_CSR_PTR XDMA_DST_STRIDE_PTR + XDMA_DST_DIM
+#define XDMA_DST_EXT_NUM 3
+#define XDMA_DST_EXT_CSR_NUM 5
+#define XDMA_DST_EXT_CUSTOM_CSR_NUM \
+    { 1, 1, 0 }
+#define XDMA_START_PTR XDMA_DST_EXT_CSR_PTR + XDMA_DST_EXT_CSR_NUM
+#define XDMA_COMMIT_TASK_PTR XDMA_START_PTR + 1
+#define XDMA_FINISH_TASK_PTR XDMA_COMMIT_TASK_PTR + 1
```
Comment (rgantonio) [inline target/snitch_cluster/sw/snax/xdma/include/snax-xdma-csr-addr.h:32]: Nice!

File: target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+            write_csr(989, value);
+            break;
+        case 990:
+            write_csr(990, value);
+            break;
+        case 991:
+            write_csr(991, value);
+            break;
+    }
+}
```
Comment (rgantonio) [inline target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c:198]: This is actually a very smart mechanic! Maybe we can write like a general function that does this too! It's cool because this time, we can put ANY value into the function!! @xiaoling-yi  can re-use this a lot and simplify most of her code.

I would very much like you add a separate library for this functions! This is so smart!

File: util/wrappergen/wrappergen.py [hunk 11]
```diff
@@ -305,6 +308,43 @@ def main():
...
+            " --readerDimension " + str(snax_xdma_cfg["reader_agu_dimension"]) +
+            " --writerDimension " + str(snax_xdma_cfg["writer_agu_dimension"]) +
+            " --readerBufferDepth " + str(snax_xdma_cfg["reader_buffer"]) +
+            " --writerBufferDepth " + str(snax_xdma_cfg["writer_buffer"]) +
+            (" --HasMemset " if snax_xdma_cfg["has_memset"] else "") +
+            (" --HasMaxPool " if snax_xdma_cfg["has_maxpool"] else "") +
+            (" --HasTransposer " if snax_xdma_cfg["has_transposer"] else "") +
+            " --target-dir " + args.gen_path +
+            cfg["cluster"]["name"] + "_xdma/"
+        )
```
Comment (rgantonio) [inline util/wrappergen/wrappergen.py:346]: This is smart too because rather than having a scala parameter generated, it can be this one. However, this implies it is very custom to the xDMA only. Not something that is quite "generalized" like the accelerators. I guess it is fine for now since the intent is to have the xDMA a SNAX thing.  But what about the different extensions? That implies that people have to (1) change this wrapper gen to add another extension, (2) modify the scala files for the extension, and (3) add the configurations too.

The SNAX accelerators just need the configuration files since they are "generalized" already and do not need too many modifications. Even the accelerator wrapper names are all centralized in the configuration files.

**Review by rgantonio — COMMENTED** [review]
Suuuupper good work @IveanEx !

I bet you work so hard on this. I like that we have this interesting xDMA already and it's a good work in progress!

I have things that need to be critical about though:
1. The PR is huge haha! It's too big honestly and very hard to read through carefully. Remember, since SNAX is something that is shared amongst our team, it is best to make "byte-sized" (get the pun? 😆 ) PRs. I will strongly recommend to keep it to less than a thousand lines. What are ways to make it small? Separate the hardware build and the software tests in separate PRs! That would be a great fix already!
2. I am a bit iffy about the differences between the streamers and the xDMA. They have the same units like the AGUs and data movers. Is there no way that they could be re-used or placed in the common cells section? You have a CommonCells directory inside the xDMA but then there are also re-usable cells in others. Maybe better to have all common cells at some directory that can be used through all! It now feels like we are not embracing the "reusability" of these modules.
3. I guess it's for testing purposes but are we really making the xDMA like a separate accelerator? I was thinking it should be inside the cluster too! My worry with this setup is that there will be a HUGE clean-up afterwards. Which I fear may be prone to more work 😅 

There are nice things I like though! Like the smart case-statements for the reading and writing for the CSRs. I hope the compiler is smart enough just to route the values! I would really love to have a separate PR actually fix that 😄 

GOOD JOB!

I won't approve yet as we need to discuss more. The next step after this is to create documentation for this! Don't under estimate it. Lack of sufficient documentation is a problem in open source stuff since we're always rushing to implement things but never taking the time to make it easier for other people to understand...

I also suggest to ask @xiaoling-yi's opinion here since there is much similarity in your modules.

**Comment by jorendumoulin** [comment]
> I guess it's for testing purposes but are we really making the xDMA like a separate accelerator? I was thinking it should be inside the cluster too! My worry with this setup is that there will be a HUGE clean-up afterwards. Which I fear may be prone to more work 😅

I'm not completely in the loop but I agree with this sentiment! There are starting to be a lot of variations on the system with the gemm accelerator, and this would add an extra one. Can we not include this xDMA in every snax cluster from now on and replace the old DMA entirely? Or are there still some disadvantages?

There would just need to be a new definition of `snitch_dma_1d` and `snitch_dma_2d` etc.., but then all of the existing tests with the dma engine will now also test your new one, so you don't have to write tests yourself 😄

File: hw/chisel/src/main/scala/snax/xdma/xdmaExtension/MaxPool.scala
[the remark this answers is not in the document]
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaExtension/MaxPool.scala:1]: The maxpool and transposer will be the replacement of @xiaoling-yi 's version. These two extensions are indeed the same as the original one.

File: hw/chisel/src/main/scala/snax/xdma/xdmaExtension/Memset.scala
[the remark this answers is not in the document]
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaExtension/Memset.scala:1]: Yes! It's the hardware implementation of POSIX's memset. The most common usage is for memory initialization, without the need of pulp's zeromem hardware.

File: hw/chisel/src/main/scala/snax/xdma/xdmaExtension/Transposer.scala
[the remark this answers is not in the document]
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaExtension/Transposer.scala:1]: The xdma extension requires the special base module, thus it is not plug-and-play.... But migration cost can be very low! 

Later, **chisel_acc** will be merged into **chisel**, as one package.

File: hw/chisel/src/main/scala/snax/xdma/xdmaTop/xdmaTop.scala [hunk 12]
```diff
@@ -138,15 +141,97 @@ class xdmaTop(
...
+    addressBufferDepth = parsed_args("writerBufferDepth").toInt
+  )
+
+  var extensionparam = Seq[HasDMAExtension]()
+  if (parsed_args.contains("HasMemset"))
+    extensionparam = extensionparam :+ HasMemset
+  if (parsed_args.contains("HasMaxPool"))
+    extensionparam = extensionparam :+ HasMaxPool
+  if (parsed_args.contains("HasTransposer"))
+    extensionparam = extensionparam :+ HasTransposer
```
[the remark this answers is not in the document]
Comment (IveanEx) [reply hw/chisel/src/main/scala/snax/xdma/xdmaTop/xdmaTop.scala:226]: Yes. The xdma extension is designed to be concatenate one with another, forming the process chain.

File: hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl (hunk 1 above)
Comment (IveanEx) [reply hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl:428]: Line 425 removed.

File: hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl (hunk 3 above)
Comment (IveanEx) [reply hw/snitch_cluster/src/snitch_cluster_wrapper.sv.tpl:905]: This is our first version, which only supports local loopback. Later this part will go inside snitch cluster and replace the xdma, and should not be generated by mako script but systemverilog's syntax.

File: target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c (hunk 5 above)
Comment (IveanEx) [reply target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c:27]: xdma will finally replace idma, locating at the last. Now the development is not finished so idma cannot be removed.

File: hw/templates/snax_xdma_wrapper.sv.tpl
Comment (IveanEx) [reply hw/templates/snax_xdma_wrapper.sv.tpl:1]: Yes, I totally agree with you! It will be done in the future updates!

File: target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c (hunk 7 above)
Comment (IveanEx) [reply target/snitch_cluster/sw/apps/snax-xdma-memset/src/snax-xdma-memset.c:30]: Modified the code. Early return if there is failure in configuring xdma.

File: target/snitch_cluster/sw/snax/xdma/include/snax-xdma-csr-addr.h (hunk 9 above)
Comment (IveanEx) [reply target/snitch_cluster/sw/snax/xdma/include/snax-xdma-csr-addr.h:32]: This part will be generated by chisel, but not implemented yet. It will be done in the next PR.

File: target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c (hunk 10 above)
Comment (IveanEx) [reply target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c:198]: Confirmed that compiler do the necessary optimization as what we wish. I will move it to the common library in the next PR.

File: util/wrappergen/wrappergen.py (hunk 11 above)
Comment (IveanEx) [reply util/wrappergen/wrappergen.py:346]: Indeed. it needs to be optimized. It should be done in a seperate PR, I think it is better to rewrite gen_chisel_file function.

**Comment by IveanEx** [comment]
> > I guess it's for testing purposes but are we really making the xDMA like a separate accelerator? I was thinking it should be inside the cluster too! My worry with this setup is that there will be a HUGE clean-up afterwards. Which I fear may be prone to more work 😅
> 
> I'm not completely in the loop but I agree with this sentiment! There are starting to be a lot of variations to the system with the gemm accelerator, and this would add an extra one. Can we not include this xDMA in every snax cluster from now on and replace the old DMA entirely? Or are there still some disadvantages?
> 
> There would just need to be a new definition of `snitch_dma_1d` and `snitch_dma_2d` etc.., but then all of the existing tests with the dma engine will now also test your new one, so you don't have to write tests yourself 😄

It cannot replace the original xdma at present, as it does not have the function to move the data from / to DRAM yet. 

Yes, like any other accelerator, it can be disabled, and which extension should be included is completely configurable. 

I also agree that now there are too many versions in snax... I plan to clean the cfg file and merge many things with @xiaoling-yi's contribution. But should happen when she come back 😊😊

File: target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c (hunk 10 above)
Comment (rgantonio) [reply target/snitch_cluster/sw/snax/xdma/src/snax-xdma-lib.c:198]: YAAAAY!

**Comment by rgantonio** [comment]
@xiaoling-yi we will push this since it's a WIP that is being cut into multiple PRs anyway. So kindly take a look when you have the time. If you see something, tell @IveanEx !

Thankkkss~

**Review by rgantonio — APPROVED** [review]
Good job!!! ❤️ ❤️ ❤️ ❤️ ❤️ 

Can you make the issue list of things to fix? Add the `- [ ]` check boxes 😄 Thankkkss!

**Comment by jorendumoulin** [comment]
> It cannot replace the original idma at present, as it does not have the function to move the data from / to DRAM yet.

Is it the plan to do this?

File: .github/workflows/gen.yml
Comment (jorendumoulin) [inline .github/workflows/gen.yml:1]: Why was this removed?

**Comment by IveanEx** [comment]
> > It cannot replace the original idma at present, as it does not have the function to move the data from / to DRAM yet.
> 
> Is it the plan to do this?

Yes, it will be done in the future. But more discussion should be made on improvement / change of AXI bus.

File: .github/workflows/gen.yml
Comment (IveanEx) [reply .github/workflows/gen.yml:1]: This is not necessary anymore. All ci use this chiselgen to generate the code, so it is naturally tested. 

Furthermore, this is the source of error that reandomly fails.

Comment (jorendumoulin) [reply .github/workflows/gen.yml:1]: Okay!
